src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/utils.py

```
@@ -27,180 +184,556 @@ def split_deepstack_embs(
+    if image_input_mask is None:
+        image_input_mask = input_ids == image_token_id
+
+    vision_values = torch.cat(vision_values)
+    vision_grid_thw = torch.tensor(vision_grid_thw, device=image_grid_thw.device, dtype=image_grid_thw.dtype)
+    vision_input_mask = video_input_mask | image_input_mask
+
+    return vision_values, vision_grid_thw, vision_input_mask
+
+
+# reference: megatron/training/utils.py get_batch_on_this_cp_rank
+def split_data_cp_rank(val: torch.Tensor, cp_size: int, seq_dim: int, cp_rank: int = None):
```

Ruff (0.15.0)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)